torch/_inductor/triton_bundler.py

```diff
@@ -0,0 +1,188 @@
+import dataclasses
+import logging
+import os
+import shutil
+import tempfile
+from pathlib import Path
+from typing import List, Optional, Tuple
+
+from torch._dynamo.utils import counters, dynamo_timed
+from torch._utils_internal import justknobs_check
+from torch.utils._ordered_set import OrderedSet
+
+from .runtime.runtime_utils import cache_dir
+
+
+log = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass(frozen=True)
+class TritonBundleEntry:
+    """
+    When we have compiled a triton kernel, we take note of that kernel by
+    its triton generated hash, its device, and where this kernel is located.
+    This is the minimum information we can use to later retreive this kernel
+    from file system.
+    """
+
+    kernel_hash: str
+    compile_time_ns: int
+    device: int
+    directory: str
+
+
+@dataclasses.dataclass(frozen=True)
+class TritonKernelArtifact:
+    """
+    Artifact for an individual kernel converted to bytes.
+    Bytes could be a cubin, json, ttir, or ttgir.
+    """
+
+    filename: str
+    payload: bytes = dataclasses.field(repr=False)  # Do not display binary
+
+
+@dataclasses.dataclass(frozen=True)
+class TritonKernelArtifacts:
+    """
+    Collection of artifacts for a particular kernel.
+    """
+
+    kernel_hash: str
+    compile_time_ns: int
+    device: int
+    artifacts: List[TritonKernelArtifact]
+
+
+@dataclasses.dataclass(frozen=True)
+class TritonBundlerMetadata:
+    """
+    Metadata used for instrumentation
+    """
+
+    num_kernels: int
+    total_compile_time_ns: int
+
+
+class TritonBundler:
+    """
+    Lightweight Triton Kernel bundler that notes each time we compile a triton
+    kernel. When collect is called, converts all the previously noted kernels and
+    their artifacts into a structured bytes blob, and later when write is called
+    it writes this structured blob back to file system.
+    """
+
+    # It is safe for multiple threads to insert entries as insert operation
+    # of OrderedSet which uses dict underhood is atomic
+    _entries: Optional[OrderedSet[TritonBundleEntry]] = None
+
+    @staticmethod
+    def is_enabled() -> bool:
+        from torch._inductor import config
+
+        if (b := config.bundle_triton_into_fx_graph_cache) is not None:
+            return b
```

Comment:
Don't you have to respect disable all caches config too?

Reply:
If inductor cache is disabled, this code will never run, so we dont need to check here

Reply:
This doesn't seem obviously true? begin_compile is called from graph.py and will then hit this, and I'm pretty sure that code gets called always